### PR TITLE
MiniAOD: change motherRef in packed Candidate

### PR DESCRIPTION
**Pull request by arizzi**
With this the reference to prunedParticles contained in the packedGen is set to the same-particle-in-pruned when available.
The behaviour of the accessors is the same and should be backward compatible.

This will allow a better merging of pruned+packed in downstream processing (e.g. for Rivet ParticleLevel producer) and avoid exact matching on pt  as done in:
https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/RivetInterface/plugins/MergedGenParticleProducer.cc#L44

@gpetruc @perrozzi @intrepid42

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: DataFormats/PatCandidates/interface/PackedGenParticle.h [hunk 1]
```diff
@@ -100,121 +100,132 @@ namespace pat {
...
-    virtual const reco::Candidate * mother( size_type ) const;
+    const reco::Candidate * mother( size_type ) const override;
     /// direct access to the mother reference (may be null)
-    const reco::GenParticleRef & motherRef() const { return mother_; }
+    reco::GenParticleRef motherRef() const { 
+	if(mother_.isNonnull() && mother_.isAvailable()&& mother_->status()==1 ){ //if pointing to the pruned version of myself
+   	  if(mother_->numberOfMothers() > 0) 
+	  	  return mother_->motherRef(0); // return my mother's (that is actually myself) mother
+	 	else
+		  return edm::Ref<reco::GenParticleCollection>(); // return null ref
```
Comment (slava77) [inline DataFormats/PatCandidates/interface/PackedGenParticle.h]: with a dummy ```static const edm::Ref<reco::GenParticleCollection> nullGPR```
this method can retain it's "return by const ref" signature

Comment (arizzi) [reply DataFormats/PatCandidates/interface/PackedGenParticle.h]: nope... the problem is in CompositePtrCandidate where motherRef(i) returns by value, not really in the nullRef(that I could handle the way you say)

Comment (arizzi) [reply DataFormats/PatCandidates/interface/PackedGenParticle.h]: I meant CompositeRefCandidate... anyhow is in the call mother_->motherRef(0)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20656	6ab622e
- changes are in line with the PR description; not affecting current tested use cases
- jenkins tests pass and comparisons show no differences

**Comment by davidlange6** [comment]
merge

**Pull request merged**